fixedbugs/issue9691.go: Compiler panic due to unhandled case for *types.Map.

This is the [fourth](https://github.com/gopherjs/gopherjs/blob/67b6a9773dd8e6c85f7c56124e9ac110d6999f32/tests/run.go#L55-L61) of four panics found via Go compiler tests added in #296.

This is the failing test case:

```Go
package main

func main() {
	s := "foo"
	b := []byte(s)
	m := make(map[string]int)
	// Test that map index can be used in range
	// and that slicebytetostringtmp is not used in this context.
	for m[string(b)] = range s {
	}
	b[0] = 'b'
	if m["foo"] != 2 {
		panic("bad")
	}
}
```

It causes a panic in `github.com/gopherjs/gopherjs/compiler` package, `(*funcContext) translateAssign` method, in [`compiler/statements.go:785`](https://github.com/gopherjs/gopherjs/blob/67b6a9773dd8e6c85f7c56124e9ac110d6999f32/compiler/statements.go#L785).

The panic is:

```
panic: Unhandled lhs type: *types.Map
```

The fix is to add `case *types.Map:` and handle it correctly.

I have attempted to solve it, but it was not trivial. It would require careful understanding of how to map index can be used in a for range statement. @neelance, your expertise there would be very helpful.

#### Reproduce

```bash
$ cd $GOPATH/src/github.com/gopherjs/gopherjs/tests
$ go run run.go -v -summary -show_known_fails -- fixedbugs/issue9691.go
goos: "darwin", goarch: "js"
# go run run.go -show_known_fails -- fixedbugs/issue9691.go
exit status 2
panic: Unhandled lhs type: *types.Map


goroutine 1 [running]:
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateAssign(0xc8201682c0, 0xe403b0, 0xc8201229c0, 0x632d38, 0x2, 0xe04688, 0x909660, 0x0, 0x0, 0x0)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/statements.go:785 +0x1931
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateStmt.func7()
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/statements.go:153 +0x370
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateLoopingStmt.func2()
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/statements.go:698 +0x162
github.com/gopherjs/gopherjs/compiler.(*funcContext).Indent(0xc8201682c0, 0xc820181250)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/utils.go:56 +0x2d
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateLoopingStmt(0xc8201682c0, 0xc820182de0, 0xc820122a20, 0xc820182ed8, 0xc8201237a0, 0x0, 0xc820130900)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/statements.go:713 +0x4a5
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateStmt(0xc8201682c0, 0xe40420, 0xc82012c9b0, 0x0)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/statements.go:160 +0x9a02
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateStmtList(0xc8201682c0, 0xc82013a700, 0x6, 0x8)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/statements.go:24 +0x9e
github.com/gopherjs/gopherjs/compiler.translateFunction.func1()
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/package.go:679 +0x9ba
github.com/gopherjs/gopherjs/compiler.(*funcContext).CatchOutput(0xc8201682c0, 0x1, 0xc820183820, 0x0, 0x0, 0x0)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/utils.go:64 +0x84
github.com/gopherjs/gopherjs/compiler.translateFunction(0xc820130880, 0x0, 0x0, 0x0, 0xc820122b70, 0xc820168160, 0xc820122d50, 0xc8201367e0, 0xc82012a600, 0x4, ...)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/package.go:680 +0x6d0
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateToplevelFunction.func1(0xc82012a600, 0x4, 0x0, 0x0, 0x0)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/package.go:562 +0x90a
github.com/gopherjs/gopherjs/compiler.(*funcContext).translateToplevelFunction(0xc820168160, 0xc820122ba0, 0xc8201367e0, 0x0, 0x0, 0x0)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/package.go:571 +0x3e2
github.com/gopherjs/gopherjs/compiler.Compile.func9()
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/package.go:340 +0x42
github.com/gopherjs/gopherjs/compiler.Compile.func5(0xc820123620, 0x0, 0x0, 0x0)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/package.go:230 +0xa6
github.com/gopherjs/gopherjs/compiler.Compile(0x6358d8, 0x4, 0xc82013c048, 0x1, 0x1, 0xc820134080, 0xc82012a490, 0x0, 0x52128f16, 0x0, ...)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/compiler/package.go:341 +0x47ec
github.com/gopherjs/gopherjs/build.(*Session).BuildPackage(0xc820130300, 0xc82012c640, 0x0, 0x0)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/build/build.go:466 +0xfbf
github.com/gopherjs/gopherjs/build.(*Session).BuildFiles(0xc820130300, 0xc82012a430, 0x1, 0x1, 0xc82012c5a0, 0x47, 0xc820130080, 0x12, 0x0, 0x0)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/build/build.go:341 +0x2c9
main.main.func3.1(0x0, 0x0)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/tool.go:235 +0x473
main.handleError(0xc820185b40, 0xc820134000, 0x0, 0x0)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/tool.go:610 +0x46
main.main.func3(0xc820138540, 0xc82012a430, 0x1, 0x1)
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/tool.go:242 +0x5c
github.com/spf13/cobra.(*Command).execute(0xc820138540, 0xc82012a3c0, 0x1, 0x1, 0x0, 0x0)
	/Users/Dmitri/Dropbox/Work/2013/GoLanding/src/github.com/spf13/cobra/command.go:499 +0x869
github.com/spf13/cobra.(*Command).Execute(0xc820138c40, 0x0, 0x0)
	/Users/Dmitri/Dropbox/Work/2013/GoLanding/src/github.com/spf13/cobra/command.go:582 +0x46a
main.main()
	/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/gopherjs/gopherjs/tool.go:461 +0x13cd

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

knfl	fixedbugs/issue9691.go	0.016s
    1 knfl
```

--- a/compiler/analysis/info.go
+++ b/compiler/analysis/info.go
@@ -208,7 +208,7 @@ func (c *FuncInfo) Visit(node ast.Node) ast.Visitor {
 			c.markBlocking(c.analyzeStack)
 		}
 	case *ast.RangeStmt:
-		if _, ok := c.p.Types[n.X].Type.Underlying().(*types.Chan); ok {
+		if _, ok := c.p.TypeOf(n.X).Underlying().(*types.Chan); ok {
 			c.markBlocking(c.analyzeStack)
 		}
 	case *ast.SelectStmt:

--- a/compiler/analysis/sideeffect.go
+++ b/compiler/analysis/sideeffect.go
@@ -24,7 +24,7 @@ func (v *hasSideEffectVisitor) Visit(node ast.Node) (w ast.Visitor) {
 	}
 	switch n := node.(type) {
 	case *ast.CallExpr:
-		if _, isSig := v.info.Types[n.Fun].Type.(*types.Signature); isSig { // skip conversions
+		if _, isSig := v.info.TypeOf(n.Fun).(*types.Signature); isSig { // skip conversions
 			v.hasSideEffect = true
 			return nil
 		}

--- a/compiler/expressions.go
+++ b/compiler/expressions.go
@@ -1094,9 +1094,6 @@ func (c *funcContext) translateImplicitConversion(expr ast.Expr, desiredType typ
 	if desiredType == nil {
 		return c.translateExpr(expr)
 	}
-	if expr == nil {
-		return c.formatExpr("%e", c.zeroValue(desiredType))
-	}
 
 	exprType := c.p.TypeOf(expr)
 	if types.Identical(exprType, desiredType) {

--- a/compiler/expressions.go
+++ b/compiler/expressions.go
@@ -108,7 +108,7 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 		collectIndexedElements := func(elementType types.Type) []string {
 			var elements []string
 			i := 0
-			zero := c.zeroValue(elementType)
+			zero := c.translateExpr(c.zeroValue(elementType)).String()
 			for _, element := range e.Elts {
 				if kve, isKve := element.(*ast.KeyValueExpr); isKve {
 					key, ok := exact.Int64Val(c.p.Types[kve.Key].Value)
@@ -131,9 +131,9 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 		case *types.Array:
 			elements := collectIndexedElements(t.Elem())
 			if len(elements) == 0 {
-				return c.formatExpr("%s", c.zeroValue(t))
+				return c.formatExpr("%s.zero()", c.typeName(t))
 			}
-			zero := c.zeroValue(t.Elem())
+			zero := c.translateExpr(c.zeroValue(t.Elem())).String()
 			for len(elements) < int(t.Len()) {
 				elements = append(elements, zero)
 			}
@@ -162,7 +162,7 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 			}
 			if isKeyValue {
 				for i := range elements {
-					elements[i] = c.zeroValue(t.Field(i).Type())
+					elements[i] = c.translateExpr(c.zeroValue(t.Field(i).Type())).String()
 				}
 				for _, element := range e.Elts {
 					kve := element.(*ast.KeyValueExpr)
@@ -462,9 +462,9 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 			}
 			key := fmt.Sprintf("%s.keyFor(%s)", c.typeName(t.Key()), c.translateImplicitConversion(e.Index, t.Key()))
 			if _, isTuple := exprType.(*types.Tuple); isTuple {
-				return c.formatExpr(`(%1s = %2e[%3s], %1s !== undefined ? [%1s.v, true] : [%4s, false])`, c.newVariable("_entry"), e.X, key, c.zeroValue(t.Elem()))
+				return c.formatExpr(`(%1s = %2e[%3s], %1s !== undefined ? [%1s.v, true] : [%4e, false])`, c.newVariable("_entry"), e.X, key, c.zeroValue(t.Elem()))
 			}
-			return c.formatExpr(`(%1s = %2e[%3s], %1s !== undefined ? %1s.v : %4s)`, c.newVariable("_entry"), e.X, key, c.zeroValue(t.Elem()))
+			return c.formatExpr(`(%1s = %2e[%3s], %1s !== undefined ? %1s.v : %4e)`, c.newVariable("_entry"), e.X, key, c.zeroValue(t.Elem()))
 		case *types.Basic:
 			return c.formatExpr("%e.charCodeAt(%f)", e.X, e.Index)
 		default:
@@ -728,7 +728,26 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 		case *types.TypeName:
 			return c.formatExpr("%s", c.typeName(o.Type()))
 		case *types.Nil:
-			return c.formatExpr("%s", c.zeroValue(c.p.TypeOf(e)))
+			if typesutil.IsJsObject(exprType) {
+				return c.formatExpr("null")
+			}
+			switch t := exprType.Underlying().(type) {
+			case *types.Basic:
+				if t.Kind() != types.UnsafePointer {
+					panic("unexpected basic type")
+				}
+				return c.formatExpr("0")
+			case *types.Slice, *types.Pointer, *types.Chan:
+				return c.formatExpr("%s.nil", c.typeName(exprType))
+			case *types.Map:
+				return c.formatExpr("false")
+			case *types.Interface:
+				return c.formatExpr("$ifaceNil")
+			case *types.Signature:
+				return c.formatExpr("$throwNilPointerError")
+			default:
+				panic(fmt.Sprintf("unexpected type: %T", t))
+			}
 		default:
 			panic(fmt.Sprintf("Unhandled object: %T\n", o))
 		}
@@ -808,9 +827,9 @@ func (c *funcContext) translateBuiltin(name string, sig *types.Signature, args [
 		}
 		switch t.Elem().Underlying().(type) {
 		case *types.Struct, *types.Array:
-			return c.formatExpr("%s", c.zeroValue(t.Elem()))
+			return c.formatExpr("%e", c.zeroValue(t.Elem()))
 		default:
-			return c.formatExpr("$newDataPointer(%s, %s)", c.zeroValue(t.Elem()), c.typeName(t))
+			return c.formatExpr("$newDataPointer(%e, %s)", c.zeroValue(t.Elem()), c.typeName(t))
 		}
 	case "make":
 		switch argType := c.p.TypeOf(args[0]).Underlying().(type) {
@@ -1035,7 +1054,7 @@ func (c *funcContext) translateConversion(expr ast.Expr, desiredType types.Type)
 			if c.p.Pkg.Path() == "syscall" && types.Identical(exprType, types.Typ[types.UnsafePointer]) {
 				array := c.newVariable("_array")
 				target := c.newVariable("_struct")
-				return c.formatExpr("(%s = %e, %s = %s, %s, %s)", array, expr, target, c.zeroValue(t.Elem()), c.loadStruct(array, target, s), target)
+				return c.formatExpr("(%s = %e, %s = %e, %s, %s)", array, expr, target, c.zeroValue(t.Elem()), c.loadStruct(array, target, s), target)
 			}
 			return c.formatExpr("$pointerOfStructConversion(%e, %s)", expr, c.typeName(t))
 		}
@@ -1076,7 +1095,7 @@ func (c *funcContext) translateImplicitConversion(expr ast.Expr, desiredType typ
 		return c.translateExpr(expr)
 	}
 	if expr == nil {
-		return c.formatExpr("%s", c.zeroValue(desiredType))
+		return c.formatExpr("%e", c.zeroValue(desiredType))
 	}
 
 	exprType := c.p.TypeOf(expr)
@@ -1086,7 +1105,7 @@ func (c *funcContext) translateImplicitConversion(expr ast.Expr, desiredType typ
 
 	basicExprType, isBasicExpr := exprType.Underlying().(*types.Basic)
 	if isBasicExpr && basicExprType.Kind() == types.UntypedNil {
-		return c.formatExpr("%s", c.zeroValue(desiredType))
+		return c.formatExpr("%e", c.zeroValue(desiredType))
 	}
 
 	switch desiredType.Underlying().(type) {

--- a/compiler/expressions.go
+++ b/compiler/expressions.go
@@ -213,17 +213,18 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 				if c.p.escapingVars[obj] {
 					return c.formatExpr("(%1s.$ptr || (%1s.$ptr = new %2s(function() { return this.$target[0]; }, function($v) { this.$target[0] = $v; }, %1s)))", c.p.objectNames[obj], c.typeName(exprType))
 				}
-				return c.formatExpr(`(%1s || (%1s = new %2s(function() { return %3s; }, function($v) { %4s })))`, c.varPtrName(obj), c.typeName(exprType), c.objectName(obj), c.translateAssign(x, "$v", exprType, false))
+				return c.formatExpr(`(%1s || (%1s = new %2s(function() { return %3s; }, function($v) { %4s })))`, c.varPtrName(obj), c.typeName(exprType), c.objectName(obj), c.translateAssign(x, c.newIdent("$v", exprType), false))
 			case *ast.SelectorExpr:
 				sel, ok := c.p.Selections[x]
 				if !ok {
 					// qualified identifier
 					obj := c.p.Uses[x.Sel].(*types.Var)
-					return c.formatExpr(`(%1s || (%1s = new %2s(function() { return %3s; }, function($v) { %4s })))`, c.varPtrName(obj), c.typeName(exprType), c.objectName(obj), c.translateAssign(x, "$v", exprType, false))
+					return c.formatExpr(`(%1s || (%1s = new %2s(function() { return %3s; }, function($v) { %4s })))`, c.varPtrName(obj), c.typeName(exprType), c.objectName(obj), c.translateAssign(x, c.newIdent("$v", exprType), false))
 				}
 				newSel := &ast.SelectorExpr{X: c.newIdent("this.$target", c.p.TypeOf(x.X)), Sel: x.Sel}
+				c.setType(newSel, exprType)
 				c.p.Selections[newSel] = sel
-				return c.formatExpr("(%1e.$ptr_%2s || (%1e.$ptr_%2s = new %3s(function() { return %4e; }, function($v) { %5s }, %1e)))", x.X, x.Sel.Name, c.typeName(exprType), newSel, c.translateAssign(newSel, "$v", exprType, false))
+				return c.formatExpr("(%1e.$ptr_%2s || (%1e.$ptr_%2s = new %3s(function() { return %4e; }, function($v) { %5s }, %1e)))", x.X, x.Sel.Name, c.typeName(exprType), newSel, c.translateAssign(newSel, c.newIdent("$v", exprType), false))
 			case *ast.IndexExpr:
 				if _, ok := c.p.TypeOf(x.X).Underlying().(*types.Slice); ok {
 					return c.formatExpr("$indexPtr(%1e.$array, %1e.$offset + %2e, %3s)", x.X, x.Index, c.typeName(exprType))

--- a/compiler/expressions.go
+++ b/compiler/expressions.go
@@ -34,7 +34,7 @@ func (e *expression) StringWithParens() string {
 }
 
 func (c *funcContext) translateExpr(expr ast.Expr) *expression {
-	exprType := c.p.Types[expr].Type
+	exprType := c.p.TypeOf(expr)
 	if value := c.p.Types[expr].Value; value != nil {
 		basic := exprType.Underlying().(*types.Basic)
 		switch {
@@ -193,7 +193,7 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 		return c.formatExpr("(%s)", fun)
 
 	case *ast.UnaryExpr:
-		t := c.p.Types[e.X].Type
+		t := c.p.TypeOf(e.X)
 		switch e.Op {
 		case token.AND:
 			if typesutil.IsJsObject(exprType) {
@@ -207,7 +207,7 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 
 			switch x := astutil.RemoveParens(e.X).(type) {
 			case *ast.CompositeLit:
-				return c.formatExpr("$newDataPointer(%e, %s)", x, c.typeName(c.p.Types[e].Type))
+				return c.formatExpr("$newDataPointer(%e, %s)", x, c.typeName(c.p.TypeOf(e)))
 			case *ast.Ident:
 				obj := c.p.Uses[x].(*types.Var)
 				if c.p.escapingVars[obj] {
@@ -221,11 +221,11 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 					obj := c.p.Uses[x.Sel].(*types.Var)
 					return c.formatExpr(`(%1s || (%1s = new %2s(function() { return %3s; }, function($v) { %4s })))`, c.varPtrName(obj), c.typeName(exprType), c.objectName(obj), c.translateAssign(x, "$v", exprType, false))
 				}
-				newSel := &ast.SelectorExpr{X: c.newIdent("this.$target", c.p.Types[x.X].Type), Sel: x.Sel}
+				newSel := &ast.SelectorExpr{X: c.newIdent("this.$target", c.p.TypeOf(x.X)), Sel: x.Sel}
 				c.p.Selections[newSel] = sel
 				return c.formatExpr("(%1e.$ptr_%2s || (%1e.$ptr_%2s = new %3s(function() { return %4e; }, function($v) { %5s }, %1e)))", x.X, x.Sel.Name, c.typeName(exprType), newSel, c.translateAssign(newSel, "$v", exprType, false))
 			case *ast.IndexExpr:
-				if _, ok := c.p.Types[x.X].Type.Underlying().(*types.Slice); ok {
+				if _, ok := c.p.TypeOf(x.X).Underlying().(*types.Slice); ok {
 					return c.formatExpr("$indexPtr(%1e.$array, %1e.$offset + %2e, %3s)", x.X, x.Index, c.typeName(exprType))
 				}
 				return c.formatExpr("$indexPtr(%e, %e, %s)", x.X, x.Index, c.typeName(exprType))
@@ -282,8 +282,8 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 			}))
 		}
 
-		t := c.p.Types[e.X].Type
-		t2 := c.p.Types[e.Y].Type
+		t := c.p.TypeOf(e.X)
+		t2 := c.p.TypeOf(e.Y)
 		_, isInterface := t2.Underlying().(*types.Interface)
 		if isInterface || types.Identical(t, types.Typ[types.UntypedNil]) {
 			t = t2
@@ -447,7 +447,7 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 		return c.formatParenExpr("%e", e.X)
 
 	case *ast.IndexExpr:
-		switch t := c.p.Types[e.X].Type.Underlying().(type) {
+		switch t := c.p.TypeOf(e.X).Underlying().(type) {
 		case *types.Array, *types.Pointer:
 			pattern := rangeCheck("%1e[%2f]", c.p.Types[e.Index].Value != nil, true)
 			if _, ok := t.(*types.Pointer); ok { // check pointer for nix (attribute getter causes a panic)
@@ -457,7 +457,7 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 		case *types.Slice:
 			return c.formatExpr(rangeCheck("%1e.$array[%1e.$offset + %2f]", c.p.Types[e.Index].Value != nil, false), e.X, e.Index)
 		case *types.Map:
-			if typesutil.IsJsObject(c.p.Types[e.Index].Type) {
+			if typesutil.IsJsObject(c.p.TypeOf(e.Index)) {
 				c.p.errList = append(c.p.errList, types.Error{Fset: c.p.fileSet, Pos: e.Index.Pos(), Msg: "cannot use js.Object as map key"})
 			}
 			key := fmt.Sprintf("%s.keyFor(%s)", c.typeName(t.Key()), c.translateImplicitConversion(e.Index, t.Key()))
@@ -472,7 +472,7 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 		}
 
 	case *ast.SliceExpr:
-		if b, isBasic := c.p.Types[e.X].Type.Underlying().(*types.Basic); isBasic && isString(b) {
+		if b, isBasic := c.p.TypeOf(e.X).Underlying().(*types.Basic); isBasic && isString(b) {
 			switch {
 			case e.Low == nil && e.High == nil:
 				return c.translateExpr(e.X)
@@ -535,10 +535,10 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 		plainFun := astutil.RemoveParens(e.Fun)
 
 		if astutil.IsTypeExpr(plainFun, c.p.Info.Info) {
-			return c.formatExpr("%s", c.translateConversion(e.Args[0], c.p.Types[plainFun].Type))
+			return c.formatExpr("%s", c.translateConversion(e.Args[0], c.p.TypeOf(plainFun)))
 		}
 
-		sig := c.p.Types[plainFun].Type.Underlying().(*types.Signature)
+		sig := c.p.TypeOf(plainFun).Underlying().(*types.Signature)
 
 		switch f := plainFun.(type) {
 		case *ast.Ident:
@@ -570,7 +570,7 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 			}
 
 			externalizeExpr := func(e ast.Expr) string {
-				t := c.p.Types[e].Type
+				t := c.p.TypeOf(e)
 				if types.Identical(t, types.Typ[types.UntypedNil]) {
 					return "null"
 				}
@@ -690,11 +690,11 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 		}
 
 	case *ast.StarExpr:
-		if typesutil.IsJsObject(c.p.Types[e.X].Type) {
+		if typesutil.IsJsObject(c.p.TypeOf(e.X)) {
 			return c.formatExpr("new $jsObjectPtr(%e)", e.X)
 		}
 		if c1, isCall := e.X.(*ast.CallExpr); isCall && len(c1.Args) == 1 {
-			if c2, isCall := c1.Args[0].(*ast.CallExpr); isCall && len(c2.Args) == 1 && types.Identical(c.p.Types[c2.Fun].Type, types.Typ[types.UnsafePointer]) {
+			if c2, isCall := c1.Args[0].(*ast.CallExpr); isCall && len(c2.Args) == 1 && types.Identical(c.p.TypeOf(c2.Fun), types.Typ[types.UnsafePointer]) {
 				if unary, isUnary := c2.Args[0].(*ast.UnaryExpr); isUnary && unary.Op == token.AND {
 					return c.translateExpr(unary.X) // unsafe conversion
 				}
@@ -710,7 +710,7 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 		if e.Type == nil {
 			return c.translateExpr(e.X)
 		}
-		t := c.p.Types[e.Type].Type
+		t := c.p.TypeOf(e.Type)
 		if _, isTuple := exprType.(*types.Tuple); isTuple {
 			return c.formatExpr("$assertType(%e, %s, true)", e.X, c.typeName(t))
 		}
@@ -728,13 +728,13 @@ func (c *funcContext) translateExpr(expr ast.Expr) *expression {
 		case *types.TypeName:
 			return c.formatExpr("%s", c.typeName(o.Type()))
 		case *types.Nil:
-			return c.formatExpr("%s", c.zeroValue(c.p.Types[e].Type))
+			return c.formatExpr("%s", c.zeroValue(c.p.TypeOf(e)))
 		default:
 			panic(fmt.Sprintf("Unhandled object: %T\n", o))
 		}
 
 	case *this:
-		if isWrapped(c.p.Types[e].Type) {
+		if isWrapped(c.p.TypeOf(e)) {
 			return c.formatExpr("this.$val")
 		}
 		return c.formatExpr("this")
@@ -813,9 +813,9 @@ func (c *funcContext) translateBuiltin(name string, sig *types.Signature, args [
 			return c.formatExpr("$newDataPointer(%s, %s)", c.zeroValue(t.Elem()), c.typeName(t))
 		}
 	case "make":
-		switch argType := c.p.Types[args[0]].Type.Underlying().(type) {
+		switch argType := c.p.TypeOf(args[0]).Underlying().(type) {
 		case *types.Slice:
-			t := c.typeName(c.p.Types[args[0]].Type)
+			t := c.typeName(c.p.TypeOf(args[0]))
 			if len(args) == 3 {
 				return c.formatExpr("$makeSlice(%s, %f, %f)", t, args[1], args[2])
 			}
@@ -827,12 +827,12 @@ func (c *funcContext) translateBuiltin(name string, sig *types.Signature, args [
 			if len(args) == 2 {
 				length = c.translateExpr(args[1]).String()
 			}
-			return c.formatExpr("new %s(%s)", c.typeName(c.p.Types[args[0]].Type), length)
+			return c.formatExpr("new %s(%s)", c.typeName(c.p.TypeOf(args[0])), length)
 		default:
 			panic(fmt.Sprintf("Unhandled make type: %T\n", argType))
 		}
 	case "len":
-		switch argType := c.p.Types[args[0]].Type.Underlying().(type) {
+		switch argType := c.p.TypeOf(args[0]).Underlying().(type) {
 		case *types.Basic:
 			return c.formatExpr("%e.length", args[0])
 		case *types.Slice:
@@ -848,7 +848,7 @@ func (c *funcContext) translateBuiltin(name string, sig *types.Signature, args [
 			panic(fmt.Sprintf("Unhandled len type: %T\n", argType))
 		}
 	case "cap":
-		switch argType := c.p.Types[args[0]].Type.Underlying().(type) {
+		switch argType := c.p.TypeOf(args[0]).Underlying().(type) {
 		case *types.Slice, *types.Chan:
 			return c.formatExpr("%e.$capacity", args[0])
 		case *types.Pointer:
@@ -867,10 +867,10 @@ func (c *funcContext) translateBuiltin(name string, sig *types.Signature, args [
 		sliceType := sig.Results().At(0).Type().Underlying().(*types.Slice)
 		return c.formatExpr("$append(%e, %s)", args[0], strings.Join(c.translateExprSlice(args[1:], sliceType.Elem()), ", "))
 	case "delete":
-		keyType := c.p.Types[args[0]].Type.Underlying().(*types.Map).Key()
+		keyType := c.p.TypeOf(args[0]).Underlying().(*types.Map).Key()
 		return c.formatExpr(`delete %e[%s.keyFor(%s)]`, args[0], c.typeName(keyType), c.translateImplicitConversion(args[1], keyType))
 	case "copy":
-		if basic, isBasic := c.p.Types[args[1]].Type.Underlying().(*types.Basic); isBasic && isString(basic) {
+		if basic, isBasic := c.p.TypeOf(args[1]).Underlying().(*types.Basic); isBasic && isString(basic) {
 			return c.formatExpr("$copyString(%e, %e)", args[0], args[1])
 		}
 		return c.formatExpr("$copySlice(%e, %e)", args[0], args[1])
@@ -918,13 +918,13 @@ func (c *funcContext) translateExprSlice(exprs []ast.Expr, desiredType types.Typ
 }
 
 func (c *funcContext) translateConversion(expr ast.Expr, desiredType types.Type) *expression {
-	exprType := c.p.Types[expr].Type
+	exprType := c.p.TypeOf(expr)
 	if types.Identical(exprType, desiredType) {
 		return c.translateExpr(expr)
 	}
 
 	if c.p.Pkg.Path() == "reflect" {
-		if call, isCall := expr.(*ast.CallExpr); isCall && types.Identical(c.p.Types[call.Fun].Type, types.Typ[types.UnsafePointer]) {
+		if call, isCall := expr.(*ast.CallExpr); isCall && types.Identical(c.p.TypeOf(call.Fun), types.Typ[types.UnsafePointer]) {
 			if ptr, isPtr := desiredType.(*types.Pointer); isPtr {
 				if named, isNamed := ptr.Elem().(*types.Named); isNamed {
 					switch named.Obj().Name() {
@@ -999,7 +999,7 @@ func (c *funcContext) translateConversion(expr ast.Expr, desiredType types.Type)
 					return c.formatExpr("new Uint8Array(0)")
 				}
 			}
-			if ptr, isPtr := c.p.Types[expr].Type.(*types.Pointer); c.p.Pkg.Path() == "syscall" && isPtr {
+			if ptr, isPtr := c.p.TypeOf(expr).(*types.Pointer); c.p.Pkg.Path() == "syscall" && isPtr {
 				if s, isStruct := ptr.Elem().Underlying().(*types.Struct); isStruct {
 					array := c.newVariable("_array")
 					target := c.newVariable("_struct")
@@ -1012,7 +1012,7 @@ func (c *funcContext) translateConversion(expr ast.Expr, desiredType types.Type)
 			}
 			if call, ok := expr.(*ast.CallExpr); ok {
 				if id, ok := call.Fun.(*ast.Ident); ok && id.Name == "new" {
-					return c.formatExpr("new Uint8Array(%d)", int(sizes32.Sizeof(c.p.Types[call.Args[0]].Type)))
+					return c.formatExpr("new Uint8Array(%d)", int(sizes32.Sizeof(c.p.TypeOf(call.Args[0]))))
 				}
 			}
 		}
@@ -1079,7 +1079,7 @@ func (c *funcContext) translateImplicitConversion(expr ast.Expr, desiredType typ
 		return c.formatExpr("%s", c.zeroValue(desiredType))
 	}
 
-	exprType := c.p.Types[expr].Type
+	exprType := c.p.TypeOf(expr)
 	if types.Identical(exprType, desiredType) {
 		return c.translateExpr(expr)
 	}
@@ -1110,7 +1110,7 @@ func (c *funcContext) translateImplicitConversion(expr ast.Expr, desiredType typ
 }
 
 func (c *funcContext) translateConversionToSlice(expr ast.Expr, desiredType types.Type) *expression {
-	switch c.p.Types[expr].Type.Underlying().(type) {
+	switch c.p.TypeOf(expr).Underlying().(type) {
 	case *types.Array, *types.Pointer:
 		return c.formatExpr("new %s(%e)", c.typeName(desiredType), expr)
 	}
@@ -1287,7 +1287,7 @@ func (c *funcContext) formatExprInternal(format string, a []interface{}, parens 
 				out.WriteString(strconv.FormatInt(d, 10))
 				return
 			}
-			if is64Bit(c.p.Types[e].Type.Underlying().(*types.Basic)) {
+			if is64Bit(c.p.TypeOf(e).Underlying().(*types.Basic)) {
 				out.WriteString("$flatten64(")
 				writeExpr("")
 				out.WriteString(")")
@@ -1298,7 +1298,7 @@ func (c *funcContext) formatExprInternal(format string, a []interface{}, parens 
 			e := a[n].(ast.Expr)
 			if val := c.p.Types[e].Value; val != nil {
 				d, _ := exact.Uint64Val(val)
-				if c.p.Types[e].Type.Underlying().(*types.Basic).Kind() == types.Int64 {
+				if c.p.TypeOf(e).Underlying().(*types.Basic).Kind() == types.Int64 {
 					out.WriteString(strconv.FormatInt(int64(d)>>32, 10))
 					return
 				}

--- a/compiler/filter/assign.go
+++ b/compiler/filter/assign.go
@@ -44,7 +44,7 @@ func Assign(stmt ast.Stmt, info *analysis.Info) ast.Stmt {
 		viaTmpVars = func(expr ast.Expr, name string) ast.Expr {
 			switch e := astutil.RemoveParens(expr).(type) {
 			case *ast.IndexExpr:
-				return astutil.SetType(info.Info, info.Types[e].Type, &ast.IndexExpr{
+				return astutil.SetType(info.Info, info.TypeOf(e), &ast.IndexExpr{
 					X:     viaTmpVars(e.X, "_slice"),
 					Index: viaTmpVars(e.Index, "_index"),
 				})
@@ -60,10 +60,10 @@ func Assign(stmt ast.Stmt, info *analysis.Info) ast.Stmt {
 					Sel: e.Sel,
 				}
 				info.Selections[newSel] = sel
-				return astutil.SetType(info.Info, info.Types[e].Type, newSel)
+				return astutil.SetType(info.Info, info.TypeOf(e), newSel)
 
 			case *ast.StarExpr:
-				return astutil.SetType(info.Info, info.Types[e].Type, &ast.StarExpr{
+				return astutil.SetType(info.Info, info.TypeOf(e), &ast.StarExpr{
 					X: viaTmpVars(e.X, "_ptr"),
 				})
 
@@ -71,7 +71,7 @@ func Assign(stmt ast.Stmt, info *analysis.Info) ast.Stmt {
 				return e
 
 			default:
-				tmpVar := astutil.NewIdent(name, info.Types[e].Type, info.Info, info.Pkg)
+				tmpVar := astutil.NewIdent(name, info.TypeOf(e), info.Info, info.Pkg)
 				list = append(list, &ast.AssignStmt{
 					Lhs: []ast.Expr{tmpVar},
 					Tok: token.DEFINE,
@@ -88,10 +88,10 @@ func Assign(stmt ast.Stmt, info *analysis.Info) ast.Stmt {
 			Lhs: []ast.Expr{lhs},
 			Tok: token.ASSIGN,
 			Rhs: []ast.Expr{
-				astutil.SetType(info.Info, info.Types[s.Lhs[0]].Type, &ast.BinaryExpr{
+				astutil.SetType(info.Info, info.TypeOf(s.Lhs[0]), &ast.BinaryExpr{
 					X:  lhs,
 					Op: op,
-					Y: astutil.SetType(info.Info, info.Types[s.Rhs[0]].Type, &ast.ParenExpr{
+					Y: astutil.SetType(info.Info, info.TypeOf(s.Rhs[0]), &ast.ParenExpr{
 						X: s.Rhs[0],
 					}),
 				}),

--- a/compiler/filter/incdecstmt.go
+++ b/compiler/filter/incdecstmt.go
@@ -12,9 +12,9 @@ import (
 
 func IncDecStmt(stmt ast.Stmt, info *analysis.Info) ast.Stmt {
 	if s, ok := stmt.(*ast.IncDecStmt); ok {
-		t := info.Types[s.X].Type
+		t := info.TypeOf(s.X)
 		if iExpr, isIExpr := s.X.(*ast.IndexExpr); isIExpr {
-			switch u := info.Types[iExpr.X].Type.Underlying().(type) {
+			switch u := info.TypeOf(iExpr.X).Underlying().(type) {
 			case *types.Array:
 				t = u.Elem()
 			case *types.Slice:

--- a/compiler/package.go
+++ b/compiler/package.go
@@ -259,7 +259,7 @@ func Compile(importPath string, files []*ast.File, fileSet *token.FileSet, impor
 		}
 		if _, ok := varsWithInit[o]; !ok {
 			d.DceDeps = collectDependencies(func() {
-				d.InitCode = []byte(fmt.Sprintf("\t\t%s = %s;\n", c.objectName(o), c.zeroValue(o.Type())))
+				d.InitCode = []byte(fmt.Sprintf("\t\t%s = %s;\n", c.objectName(o), c.translateExpr(c.zeroValue(o.Type())).String()))
 			})
 		}
 		d.DceObjectFilter = o.Name()
@@ -392,7 +392,7 @@ func Compile(importPath string, files []*ast.File, fileSet *token.FileSet, impor
 					}
 					constructor = fmt.Sprintf("function(%s) {\n\t\tthis.$val = this;\n\t\tif (arguments.length === 0) {\n", strings.Join(params, ", "))
 					for i := 0; i < t.NumFields(); i++ {
-						constructor += fmt.Sprintf("\t\t\tthis.%s = %s;\n", fieldName(t, i), c.zeroValue(t.Field(i).Type()))
+						constructor += fmt.Sprintf("\t\t\tthis.%s = %s;\n", fieldName(t, i), c.translateExpr(c.zeroValue(t.Field(i).Type())).String())
 					}
 					constructor += "\t\t\treturn;\n\t\t}\n"
 					for i := 0; i < t.NumFields(); i++ {
@@ -668,7 +668,7 @@ func translateFunction(typ *ast.FuncType, initStmts []ast.Stmt, body *ast.BlockS
 			c.resultNames = make([]ast.Expr, c.sig.Results().Len())
 			for i := 0; i < c.sig.Results().Len(); i++ {
 				result := c.sig.Results().At(i)
-				c.Printf("%s = %s;", c.objectName(result), c.zeroValue(result.Type()))
+				c.Printf("%s = %s;", c.objectName(result), c.translateExpr(c.zeroValue(result.Type())).String())
 				id := ast.NewIdent("")
 				c.p.Uses[id] = result
 				c.resultNames[i] = c.setType(id, result.Type())

--- a/compiler/statements.go
+++ b/compiler/statements.go
@@ -150,10 +150,10 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 			c.translateLoopingStmt(func() string { return iVar + " < " + refVar + ".length" }, s.Body, func() {
 				c.Printf("%s = $decodeRune(%s, %s);", runeVar, refVar, iVar)
 				if !isBlank(s.Key) {
-					c.Printf("%s", c.translateAssign(s.Key, iVar, types.Typ[types.Int], s.Tok == token.DEFINE))
+					c.Printf("%s", c.translateAssign(s.Key, c.newIdent(iVar, types.Typ[types.Int]), s.Tok == token.DEFINE))
 				}
 				if !isBlank(s.Value) {
-					c.Printf("%s", c.translateAssign(s.Value, runeVar+"[0]", types.Typ[types.Rune], s.Tok == token.DEFINE))
+					c.Printf("%s", c.translateAssign(s.Value, c.newIdent(runeVar+"[0]", types.Typ[types.Rune]), s.Tok == token.DEFINE))
 				}
 			}, func() {
 				c.Printf("%s += %s[1];", iVar, runeVar)
@@ -172,10 +172,10 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 					Body: &ast.BlockStmt{List: []ast.Stmt{&ast.BranchStmt{Tok: token.CONTINUE}}},
 				}, nil)
 				if !isBlank(s.Key) {
-					c.Printf("%s", c.translateAssign(s.Key, entryVar+".k", t.Key(), s.Tok == token.DEFINE))
+					c.Printf("%s", c.translateAssign(s.Key, c.newIdent(entryVar+".k", t.Key()), s.Tok == token.DEFINE))
 				}
 				if !isBlank(s.Value) {
-					c.Printf("%s", c.translateAssign(s.Value, entryVar+".v", t.Elem(), s.Tok == token.DEFINE))
+					c.Printf("%s", c.translateAssign(s.Value, c.newIdent(entryVar+".v", t.Elem()), s.Tok == token.DEFINE))
 				}
 			}, func() {
 				c.Printf("%s++;", iVar)
@@ -199,13 +199,13 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 			c.Printf("%s = 0;", iVar)
 			c.translateLoopingStmt(func() string { return iVar + " < " + length }, s.Body, func() {
 				if !isBlank(s.Key) {
-					c.Printf("%s", c.translateAssign(s.Key, iVar, types.Typ[types.Int], s.Tok == token.DEFINE))
+					c.Printf("%s", c.translateAssign(s.Key, c.newIdent(iVar, types.Typ[types.Int]), s.Tok == token.DEFINE))
 				}
 				if !isBlank(s.Value) {
-					c.Printf("%s", c.translateAssign(s.Value, c.translateImplicitConversion(c.setType(&ast.IndexExpr{
+					c.Printf("%s", c.translateAssign(s.Value, c.setType(&ast.IndexExpr{
 						X:     c.newIdent(refVar, t),
 						Index: c.newIdent(iVar, types.Typ[types.Int]),
-					}, elemType), elemType).String(), elemType, s.Tok == token.DEFINE))
+					}, elemType), s.Tok == token.DEFINE))
 				}
 			}, func() {
 				c.Printf("%s++;", iVar)
@@ -332,7 +332,7 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 				}
 				return
 			}
-			c.Printf("%s", c.translateAssignOfExpr(lhs, s.Rhs[0], s.Tok == token.DEFINE))
+			c.Printf("%s", c.translateAssign(lhs, s.Rhs[0], s.Tok == token.DEFINE))
 
 		case len(s.Lhs) > 1 && len(s.Rhs) == 1:
 			tupleVar := c.newVariable("_tuple")
@@ -341,7 +341,7 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 			for i, lhs := range s.Lhs {
 				lhs = astutil.RemoveParens(lhs)
 				if !isBlank(lhs) {
-					out += " " + c.translateAssignOfExpr(lhs, c.newIdent(fmt.Sprintf("%s[%d]", tupleVar, i), tuple.At(i).Type()), s.Tok == token.DEFINE)
+					out += " " + c.translateAssign(lhs, c.newIdent(fmt.Sprintf("%s[%d]", tupleVar, i), tuple.At(i).Type()), s.Tok == token.DEFINE)
 				}
 			}
 			c.Printf("%s", out)
@@ -356,12 +356,12 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 					}
 					continue
 				}
-				parts = append(parts, c.translateAssignOfExpr(c.newIdent(tmpVars[i], c.p.TypeOf(s.Lhs[i])), rhs, true))
+				parts = append(parts, c.translateAssign(c.newIdent(tmpVars[i], c.p.TypeOf(s.Lhs[i])), rhs, true))
 			}
 			for i, lhs := range s.Lhs {
 				lhs = astutil.RemoveParens(lhs)
 				if !isBlank(lhs) {
-					parts = append(parts, c.translateAssignOfExpr(lhs, c.newIdent(tmpVars[i], c.p.TypeOf(lhs)), s.Tok == token.DEFINE))
+					parts = append(parts, c.translateAssign(lhs, c.newIdent(tmpVars[i], c.p.TypeOf(lhs)), s.Tok == token.DEFINE))
 				}
 			}
 			c.Printf("%s", strings.Join(parts, " "))
@@ -697,7 +697,12 @@ func (c *funcContext) translateLoopingStmt(cond func() string, body *ast.BlockSt
 	c.PrintCond(!flatten, "}", fmt.Sprintf("$s = %d; continue; case %d:", data.beginCase, data.endCase))
 }
 
-func (c *funcContext) translateAssignOfExpr(lhs, rhs ast.Expr, define bool) string {
+func (c *funcContext) translateAssign(lhs, rhs ast.Expr, define bool) string {
+	lhs = astutil.RemoveParens(lhs)
+	if isBlank(lhs) {
+		panic("translateAssign with blank lhs")
+	}
+
 	if l, ok := lhs.(*ast.IndexExpr); ok {
 		if t, ok := c.p.TypeOf(l.X).Underlying().(*types.Map); ok {
 			if typesutil.IsJsObject(c.p.TypeOf(l.Index)) {
@@ -709,29 +714,22 @@ func (c *funcContext) translateAssignOfExpr(lhs, rhs ast.Expr, define bool) stri
 	}
 
 	lhsType := c.p.TypeOf(lhs)
+	rhsExpr := c.translateImplicitConversion(rhs, lhsType)
 	if _, ok := rhs.(*ast.CompositeLit); ok && define {
-		return fmt.Sprintf("%s = %s;", c.translateExpr(lhs), c.translateImplicitConversion(rhs, lhsType)) // skip $copy
-	}
-	return c.translateAssign(lhs, c.translateImplicitConversion(rhs, lhsType).String(), lhsType, define)
-}
-
-func (c *funcContext) translateAssign(lhs ast.Expr, rhs string, typ types.Type, define bool) string {
-	lhs = astutil.RemoveParens(lhs)
-	if isBlank(lhs) {
-		panic("translateAssign with blank lhs")
+		return fmt.Sprintf("%s = %s;", c.translateExpr(lhs), rhsExpr) // skip $copy
 	}
 
 	isReflectValue := false
-	if named, ok := typ.(*types.Named); ok && named.Obj().Pkg() != nil && named.Obj().Pkg().Path() == "reflect" && named.Obj().Name() == "Value" {
+	if named, ok := lhsType.(*types.Named); ok && named.Obj().Pkg() != nil && named.Obj().Pkg().Path() == "reflect" && named.Obj().Name() == "Value" {
 		isReflectValue = true
 	}
 	if !isReflectValue { // this is a performance hack, but it is safe since reflect.Value has no exported fields and the reflect package does not violate this assumption
-		switch typ.Underlying().(type) {
+		switch lhsType.Underlying().(type) {
 		case *types.Array, *types.Struct:
 			if define {
-				return fmt.Sprintf("%s = $clone(%s, %s);", c.translateExpr(lhs), rhs, c.typeName(typ))
+				return fmt.Sprintf("%s = $clone(%s, %s);", c.translateExpr(lhs), rhsExpr, c.typeName(lhsType))
 			}
-			return fmt.Sprintf("$copy(%s, %s, %s);", c.translateExpr(lhs), rhs, c.typeName(typ))
+			return fmt.Sprintf("$copy(%s, %s, %s);", c.translateExpr(lhs), rhsExpr, c.typeName(lhsType))
 		}
 	}
 
@@ -741,20 +739,20 @@ func (c *funcContext) translateAssign(lhs ast.Expr, rhs string, typ types.Type, 
 		if o == nil {
 			o = c.p.Uses[l]
 		}
-		return fmt.Sprintf("%s = %s;", c.objectName(o), rhs)
+		return fmt.Sprintf("%s = %s;", c.objectName(o), rhsExpr)
 	case *ast.SelectorExpr:
 		sel, ok := c.p.Selections[l]
 		if !ok {
 			// qualified identifier
-			return fmt.Sprintf("%s = %s;", c.objectName(c.p.Uses[l.Sel]), rhs)
+			return fmt.Sprintf("%s = %s;", c.objectName(c.p.Uses[l.Sel]), rhsExpr)
 		}
 		fields, jsTag := c.translateSelection(sel, l.Pos())
 		if jsTag != "" {
-			return fmt.Sprintf("%s.%s.%s = %s;", c.translateExpr(l.X), strings.Join(fields, "."), jsTag, c.externalize(rhs, sel.Type()))
+			return fmt.Sprintf("%s.%s.%s = %s;", c.translateExpr(l.X), strings.Join(fields, "."), jsTag, c.externalize(rhsExpr.String(), sel.Type()))
 		}
-		return fmt.Sprintf("%s.%s = %s;", c.translateExpr(l.X), strings.Join(fields, "."), rhs)
+		return fmt.Sprintf("%s.%s = %s;", c.translateExpr(l.X), strings.Join(fields, "."), rhsExpr)
 	case *ast.StarExpr:
-		return fmt.Sprintf("%s.$set(%s);", c.translateExpr(l.X), rhs)
+		return fmt.Sprintf("%s.$set(%s);", c.translateExpr(l.X), rhsExpr)
 	case *ast.IndexExpr:
 		switch t := c.p.TypeOf(l.X).Underlying().(type) {
 		case *types.Array, *types.Pointer:
@@ -762,9 +760,9 @@ func (c *funcContext) translateAssign(lhs ast.Expr, rhs string, typ types.Type, 
 			if _, ok := t.(*types.Pointer); ok { // check pointer for nix (attribute getter causes a panic)
 				pattern = `%1e.nilCheck, ` + pattern
 			}
-			return c.formatExpr(pattern, l.X, l.Index, rhs).String() + ";"
+			return c.formatExpr(pattern, l.X, l.Index, rhsExpr).String() + ";"
 		case *types.Slice:
-			return c.formatExpr(rangeCheck("%1e.$array[%1e.$offset + %2f] = %3s", c.p.Types[l.Index].Value != nil, false), l.X, l.Index, rhs).String() + ";"
+			return c.formatExpr(rangeCheck("%1e.$array[%1e.$offset + %2f] = %3s", c.p.Types[l.Index].Value != nil, false), l.X, l.Index, rhsExpr).String() + ";"
 		default:
 			panic(fmt.Sprintf("Unhandled lhs type: %T\n", t))
 		}

--- a/compiler/statements.go
+++ b/compiler/statements.go
@@ -332,8 +332,7 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 				}
 				return
 			}
-			lhsType := c.p.TypeOf(s.Lhs[0])
-			c.Printf("%s", c.translateAssignOfExpr(lhs, s.Rhs[0], lhsType, s.Tok == token.DEFINE))
+			c.Printf("%s", c.translateAssignOfExpr(lhs, s.Rhs[0], s.Tok == token.DEFINE))
 
 		case len(s.Lhs) > 1 && len(s.Rhs) == 1:
 			tupleVar := c.newVariable("_tuple")
@@ -342,8 +341,7 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 			for i, lhs := range s.Lhs {
 				lhs = astutil.RemoveParens(lhs)
 				if !isBlank(lhs) {
-					lhsType := c.p.TypeOf(s.Lhs[i])
-					out += " " + c.translateAssignOfExpr(lhs, c.newIdent(fmt.Sprintf("%s[%d]", tupleVar, i), tuple.At(i).Type()), lhsType, s.Tok == token.DEFINE)
+					out += " " + c.translateAssignOfExpr(lhs, c.newIdent(fmt.Sprintf("%s[%d]", tupleVar, i), tuple.At(i).Type()), s.Tok == token.DEFINE)
 				}
 			}
 			c.Printf("%s", out)
@@ -358,14 +356,12 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 					}
 					continue
 				}
-				lhsType := c.p.TypeOf(s.Lhs[i])
-				parts = append(parts, c.translateAssignOfExpr(c.newIdent(tmpVars[i], c.p.TypeOf(s.Lhs[i])), rhs, lhsType, true))
+				parts = append(parts, c.translateAssignOfExpr(c.newIdent(tmpVars[i], c.p.TypeOf(s.Lhs[i])), rhs, true))
 			}
 			for i, lhs := range s.Lhs {
 				lhs = astutil.RemoveParens(lhs)
 				if !isBlank(lhs) {
-					t := c.p.TypeOf(lhs)
-					parts = append(parts, c.translateAssignOfExpr(lhs, c.newIdent(tmpVars[i], t), t, s.Tok == token.DEFINE))
+					parts = append(parts, c.translateAssignOfExpr(lhs, c.newIdent(tmpVars[i], c.p.TypeOf(lhs)), s.Tok == token.DEFINE))
 				}
 			}
 			c.Printf("%s", strings.Join(parts, " "))
@@ -701,7 +697,7 @@ func (c *funcContext) translateLoopingStmt(cond func() string, body *ast.BlockSt
 	c.PrintCond(!flatten, "}", fmt.Sprintf("$s = %d; continue; case %d:", data.beginCase, data.endCase))
 }
 
-func (c *funcContext) translateAssignOfExpr(lhs, rhs ast.Expr, typ types.Type, define bool) string {
+func (c *funcContext) translateAssignOfExpr(lhs, rhs ast.Expr, define bool) string {
 	if l, ok := lhs.(*ast.IndexExpr); ok {
 		if t, ok := c.p.TypeOf(l.X).Underlying().(*types.Map); ok {
 			if typesutil.IsJsObject(c.p.TypeOf(l.Index)) {
@@ -712,10 +708,11 @@ func (c *funcContext) translateAssignOfExpr(lhs, rhs ast.Expr, typ types.Type, d
 		}
 	}
 
+	lhsType := c.p.TypeOf(lhs)
 	if _, ok := rhs.(*ast.CompositeLit); ok && define {
-		return fmt.Sprintf("%s = %s;", c.translateExpr(lhs), c.translateImplicitConversion(rhs, typ)) // skip $copy
+		return fmt.Sprintf("%s = %s;", c.translateExpr(lhs), c.translateImplicitConversion(rhs, lhsType)) // skip $copy
 	}
-	return c.translateAssign(lhs, c.translateImplicitConversion(rhs, typ).String(), typ, define)
+	return c.translateAssign(lhs, c.translateImplicitConversion(rhs, lhsType).String(), lhsType, define)
 }
 
 func (c *funcContext) translateAssign(lhs ast.Expr, rhs string, typ types.Type, define bool) string {

--- a/compiler/statements.go
+++ b/compiler/statements.go
@@ -386,12 +386,11 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 					lhs[i] = name
 				}
 				rhs := valueSpec.Values
-				isTuple := false
-				if len(rhs) == 1 {
-					_, isTuple = c.p.TypeOf(rhs[0]).(*types.Tuple)
-				}
-				for len(rhs) < len(lhs) && !isTuple {
-					rhs = append(rhs, nil)
+				if len(rhs) == 0 {
+					rhs = make([]ast.Expr, len(lhs))
+					for i, e := range lhs {
+						rhs[i] = c.zeroValue(c.p.TypeOf(e))
+					}
 				}
 				c.translateStmt(&ast.AssignStmt{
 					Lhs: lhs,
@@ -778,28 +777,29 @@ func (c *funcContext) translateAssign(lhs ast.Expr, rhs string, typ types.Type, 
 }
 
 func (c *funcContext) translateResults(results []ast.Expr) string {
-	switch c.sig.Results().Len() {
+	tuple := c.sig.Results()
+	switch tuple.Len() {
 	case 0:
 		return ""
 	case 1:
-		var result ast.Expr
+		result := c.zeroValue(tuple.At(0).Type())
 		if results != nil {
 			result = results[0]
 		}
-		v := c.translateImplicitConversion(result, c.sig.Results().At(0).Type())
+		v := c.translateImplicitConversion(result, tuple.At(0).Type())
 		c.delayedOutput = nil
 		return " " + v.String()
 	default:
 		if len(results) == 1 {
 			return " " + c.translateExpr(results[0]).String()
 		}
-		values := make([]string, c.sig.Results().Len())
+		values := make([]string, tuple.Len())
 		for i := range values {
-			var result ast.Expr
+			result := c.zeroValue(tuple.At(i).Type())
 			if results != nil {
 				result = results[i]
 			}
-			values[i] = c.translateImplicitConversion(result, c.sig.Results().At(i).Type()).String()
+			values[i] = c.translateImplicitConversion(result, tuple.At(i).Type()).String()
 		}
 		c.delayedOutput = nil
 		return " [" + strings.Join(values, ", ") + "]"

--- a/compiler/statements.go
+++ b/compiler/statements.go
@@ -77,7 +77,7 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 		if c.p.Types[tag].Value == nil {
 			refVar := c.newVariable("_ref")
 			c.Printf("%s = %s;", refVar, c.translateExpr(tag))
-			tag = c.newIdent(refVar, c.p.Types[tag].Type)
+			tag = c.newIdent(refVar, c.p.TypeOf(tag))
 		}
 
 		translateCond := func(cond ast.Expr) *expression {
@@ -103,7 +103,7 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 				value := refVar
 				caseClause := s.Body.List[index].(*ast.CaseClause)
 				if len(caseClause.List) == 1 {
-					t := c.p.Types[caseClause.List[0]].Type
+					t := c.p.TypeOf(caseClause.List[0])
 					if _, isInterface := t.Underlying().(*types.Interface); !isInterface && !types.Identical(t, types.Typ[types.UntypedNil]) {
 						value += ".$val"
 					}
@@ -115,10 +115,10 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 		}
 		c.Printf("%s = %s;", refVar, c.translateExpr(expr))
 		translateCond := func(cond ast.Expr) *expression {
-			if types.Identical(c.p.Types[cond].Type, types.Typ[types.UntypedNil]) {
+			if types.Identical(c.p.TypeOf(cond), types.Typ[types.UntypedNil]) {
 				return c.formatExpr("%s === $ifaceNil", refVar)
 			}
-			return c.formatExpr("$assertType(%s, %s, true)[1]", refVar, c.typeName(c.p.Types[cond].Type))
+			return c.formatExpr("$assertType(%s, %s, true)[1]", refVar, c.typeName(c.p.TypeOf(cond)))
 		}
 		c.translateBranchingStmt(s.Body.List, true, translateCond, printCaseBodyPrefix, label, c.Flattened[s])
 
@@ -142,7 +142,7 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 		refVar := c.newVariable("_ref")
 		c.Printf("%s = %s;", refVar, c.translateExpr(s.X))
 
-		switch t := c.p.Types[s.X].Type.Underlying().(type) {
+		switch t := c.p.TypeOf(s.X).Underlying().(type) {
 		case *types.Basic:
 			iVar := c.newVariable("_i")
 			c.Printf("%s = 0;", iVar)
@@ -298,7 +298,7 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 		case *ast.SelectorExpr:
 			isJs = typesutil.IsJsPackage(c.p.Uses[fun.Sel].Pkg())
 		}
-		sig := c.p.Types[s.Call.Fun].Type.Underlying().(*types.Signature)
+		sig := c.p.TypeOf(s.Call.Fun).Underlying().(*types.Signature)
 		args := c.translateArgs(sig, s.Call.Args, s.Call.Ellipsis.IsValid(), true)
 		if isBuiltin || isJs {
 			vars := make([]string, len(s.Call.Args))
@@ -306,7 +306,7 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 			for i, arg := range s.Call.Args {
 				v := c.newVariable("_arg")
 				vars[i] = v
-				callArgs[i] = c.newIdent(v, c.p.Types[arg].Type)
+				callArgs[i] = c.newIdent(v, c.p.TypeOf(arg))
 			}
 			call := c.translateExpr(&ast.CallExpr{
 				Fun:      s.Call.Fun,
@@ -323,18 +323,6 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 			panic(s.Tok)
 		}
 
-		if s.Tok == token.DEFINE {
-			for _, lhs := range s.Lhs {
-				if !isBlank(lhs) {
-					obj := c.p.Defs[lhs.(*ast.Ident)]
-					if obj == nil {
-						obj = c.p.Uses[lhs.(*ast.Ident)]
-					}
-					c.setType(lhs, obj.Type())
-				}
-			}
-		}
-
 		switch {
 		case len(s.Lhs) == 1 && len(s.Rhs) == 1:
 			lhs := astutil.RemoveParens(s.Lhs[0])
@@ -344,17 +332,17 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 				}
 				return
 			}
-			lhsType := c.p.Types[s.Lhs[0]].Type
+			lhsType := c.p.TypeOf(s.Lhs[0])
 			c.Printf("%s", c.translateAssignOfExpr(lhs, s.Rhs[0], lhsType, s.Tok == token.DEFINE))
 
 		case len(s.Lhs) > 1 && len(s.Rhs) == 1:
 			tupleVar := c.newVariable("_tuple")
 			out := tupleVar + " = " + c.translateExpr(s.Rhs[0]).String() + ";"
-			tuple := c.p.Types[s.Rhs[0]].Type.(*types.Tuple)
+			tuple := c.p.TypeOf(s.Rhs[0]).(*types.Tuple)
 			for i, lhs := range s.Lhs {
 				lhs = astutil.RemoveParens(lhs)
 				if !isBlank(lhs) {
-					lhsType := c.p.Types[s.Lhs[i]].Type
+					lhsType := c.p.TypeOf(s.Lhs[i])
 					out += " " + c.translateAssignOfExpr(lhs, c.newIdent(fmt.Sprintf("%s[%d]", tupleVar, i), tuple.At(i).Type()), lhsType, s.Tok == token.DEFINE)
 				}
 			}
@@ -370,13 +358,13 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 					}
 					continue
 				}
-				lhsType := c.p.Types[s.Lhs[i]].Type
-				parts = append(parts, c.translateAssignOfExpr(c.newIdent(tmpVars[i], c.p.Types[s.Lhs[i]].Type), rhs, lhsType, true))
+				lhsType := c.p.TypeOf(s.Lhs[i])
+				parts = append(parts, c.translateAssignOfExpr(c.newIdent(tmpVars[i], c.p.TypeOf(s.Lhs[i])), rhs, lhsType, true))
 			}
 			for i, lhs := range s.Lhs {
 				lhs = astutil.RemoveParens(lhs)
 				if !isBlank(lhs) {
-					t := c.p.Types[lhs].Type
+					t := c.p.TypeOf(lhs)
 					parts = append(parts, c.translateAssignOfExpr(lhs, c.newIdent(tmpVars[i], t), t, s.Tok == token.DEFINE))
 				}
 			}
@@ -400,7 +388,7 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 				rhs := valueSpec.Values
 				isTuple := false
 				if len(rhs) == 1 {
-					_, isTuple = c.p.Types[rhs[0]].Type.(*types.Tuple)
+					_, isTuple = c.p.TypeOf(rhs[0]).(*types.Tuple)
 				}
 				for len(rhs) < len(lhs) && !isTuple {
 					rhs = append(rhs, nil)
@@ -436,10 +424,10 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 		c.translateStmt(s.Stmt, label)
 
 	case *ast.GoStmt:
-		c.Printf("$go(%s, [%s]);", c.translateExpr(s.Call.Fun), strings.Join(c.translateArgs(c.p.Types[s.Call.Fun].Type.Underlying().(*types.Signature), s.Call.Args, s.Call.Ellipsis.IsValid(), true), ", "))
+		c.Printf("$go(%s, [%s]);", c.translateExpr(s.Call.Fun), strings.Join(c.translateArgs(c.p.TypeOf(s.Call.Fun).Underlying().(*types.Signature), s.Call.Args, s.Call.Ellipsis.IsValid(), true), ", "))
 
 	case *ast.SendStmt:
-		chanType := c.p.Types[s.Chan].Type.Underlying().(*types.Chan)
+		chanType := c.p.TypeOf(s.Chan).Underlying().(*types.Chan)
 		call := &ast.CallExpr{
 			Fun:  c.newIdent("$send", types.NewSignature(nil, types.NewTuple(types.NewVar(0, nil, "", chanType), types.NewVar(0, nil, "", chanType.Elem())), nil, false)),
 			Args: []ast.Expr{s.Chan, c.newIdent(c.translateImplicitConversionWithCloning(s.Value, chanType.Elem()).String(), chanType.Elem())},
@@ -463,7 +451,7 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 			case *ast.AssignStmt:
 				channels = append(channels, c.formatExpr("[%e]", astutil.RemoveParens(comm.Rhs[0]).(*ast.UnaryExpr).X).String())
 			case *ast.SendStmt:
-				chanType := c.p.Types[comm.Chan].Type.Underlying().(*types.Chan)
+				chanType := c.p.TypeOf(comm.Chan).Underlying().(*types.Chan)
 				channels = append(channels, c.formatExpr("[%e, %s]", comm.Chan, c.translateImplicitConversionWithCloning(comm.Value, chanType.Elem())).String())
 			default:
 				panic(fmt.Sprintf("unhandled: %T", comm))
@@ -490,7 +478,7 @@ func (c *funcContext) translateStmt(stmt ast.Stmt, label *types.Label) {
 		}
 		printCaseBodyPrefix := func(index int) {
 			if assign, ok := s.Body.List[index].(*ast.CommClause).Comm.(*ast.AssignStmt); ok {
-				switch rhsType := c.p.Types[assign.Rhs[0]].Type.(type) {
+				switch rhsType := c.p.TypeOf(assign.Rhs[0]).(type) {
 				case *types.Tuple:
 					c.translateStmt(&ast.AssignStmt{Lhs: assign.Lhs, Rhs: []ast.Expr{c.newIdent(selectionVar+"[1]", rhsType)}, Tok: assign.Tok}, nil)
 				default:
@@ -716,8 +704,8 @@ func (c *funcContext) translateLoopingStmt(cond func() string, body *ast.BlockSt
 
 func (c *funcContext) translateAssignOfExpr(lhs, rhs ast.Expr, typ types.Type, define bool) string {
 	if l, ok := lhs.(*ast.IndexExpr); ok {
-		if t, ok := c.p.Types[l.X].Type.Underlying().(*types.Map); ok {
-			if typesutil.IsJsObject(c.p.Types[l.Index].Type) {
+		if t, ok := c.p.TypeOf(l.X).Underlying().(*types.Map); ok {
+			if typesutil.IsJsObject(c.p.TypeOf(l.Index)) {
 				c.p.errList = append(c.p.errList, types.Error{Fset: c.p.fileSet, Pos: l.Index.Pos(), Msg: "cannot use js.Object as map key"})
 			}
 			keyVar := c.newVariable("_key")
@@ -772,7 +760,7 @@ func (c *funcContext) translateAssign(lhs ast.Expr, rhs string, typ types.Type, 
 	case *ast.StarExpr:
 		return fmt.Sprintf("%s.$set(%s);", c.translateExpr(l.X), rhs)
 	case *ast.IndexExpr:
-		switch t := c.p.Types[l.X].Type.Underlying().(type) {
+		switch t := c.p.TypeOf(l.X).Underlying().(type) {
 		case *types.Array, *types.Pointer:
 			pattern := rangeCheck("%1e[%2f] = %3s", c.p.Types[l.Index].Value != nil, true)
 			if _, ok := t.(*types.Pointer); ok { // check pointer for nix (attribute getter causes a panic)

--- a/compiler/utils.go
+++ b/compiler/utils.go
@@ -75,7 +75,7 @@ func (c *funcContext) Delayed(f func()) {
 
 func (c *funcContext) translateArgs(sig *types.Signature, argExprs []ast.Expr, ellipsis, clone bool) []string {
 	if len(argExprs) == 1 {
-		if tuple, isTuple := c.p.Types[argExprs[0]].Type.(*types.Tuple); isTuple {
+		if tuple, isTuple := c.p.TypeOf(argExprs[0]).(*types.Tuple); isTuple {
 			tupleVar := c.newVariable("_tuple")
 			c.Printf("%s = %s;", tupleVar, c.translateExpr(argExprs[0]))
 			argExprs = make([]ast.Expr, tuple.Len())

--- a/compiler/utils.go
+++ b/compiler/utils.go
@@ -177,11 +177,19 @@ func (c *funcContext) zeroValue(ty types.Type) ast.Expr {
 			return c.newConst(ty, exact.MakeInt64(0))
 		case isString(t):
 			return c.newConst(ty, exact.MakeString(""))
+		case t.Kind() == types.UnsafePointer:
+			// fall through to "nil"
 		case t.Kind() == types.UntypedNil:
 			panic("Zero value for untyped nil.")
+		default:
+			panic(fmt.Sprintf("Unhandled basic type: %v\n", t))
 		}
 	case *types.Array, *types.Struct:
 		return c.setType(&ast.CompositeLit{}, ty)
+	case *types.Chan, *types.Interface, *types.Map, *types.Signature, *types.Slice, *types.Pointer:
+		// fall through to "nil"
+	default:
+		panic(fmt.Sprintf("Unhandled type: %T\n", t))
 	}
 	id := c.newIdent("nil", ty)
 	c.p.Uses[id] = nilObj

--- a/compiler/utils.go
+++ b/compiler/utils.go
@@ -13,6 +13,7 @@ import (
 	"github.com/gopherjs/gopherjs/compiler/analysis"
 	"github.com/gopherjs/gopherjs/compiler/typesutil"
 
+	"golang.org/x/tools/go/exact"
 	"golang.org/x/tools/go/types"
 )
 
@@ -164,41 +165,33 @@ func (c *funcContext) translateSelection(sel *types.Selection, pos token.Pos) ([
 	return fields, ""
 }
 
-func (c *funcContext) zeroValue(ty types.Type) string {
-	if typesutil.IsJsObject(ty) {
-		return "null"
-	}
+var nilObj = types.Universe.Lookup("nil")
+
+func (c *funcContext) zeroValue(ty types.Type) ast.Expr {
 	switch t := ty.Underlying().(type) {
 	case *types.Basic:
 		switch {
-		case is64Bit(t) || isComplex(t):
-			return fmt.Sprintf("new %s(0, 0)", c.typeName(ty))
 		case isBoolean(t):
-			return "false"
-		case isNumeric(t), t.Kind() == types.UnsafePointer:
-			return "0"
+			return c.newConst(ty, exact.MakeBool(false))
+		case isNumeric(t):
+			return c.newConst(ty, exact.MakeInt64(0))
 		case isString(t):
-			return `""`
+			return c.newConst(ty, exact.MakeString(""))
 		case t.Kind() == types.UntypedNil:
 			panic("Zero value for untyped nil.")
-		default:
-			panic("Unhandled type")
 		}
-	case *types.Array:
-		return fmt.Sprintf("%s.zero()", c.typeName(ty))
-	case *types.Signature:
-		return "$throwNilPointerError"
-	case *types.Slice, *types.Pointer, *types.Chan:
-		return fmt.Sprintf("%s.nil", c.typeName(ty))
-	case *types.Struct:
-		return fmt.Sprintf("new %s.ptr()", c.typeName(ty))
-	case *types.Map:
-		return "false"
-	case *types.Interface:
-		return "$ifaceNil"
-	default:
-		panic(fmt.Sprintf("unexpected type: %T", t))
+	case *types.Array, *types.Struct:
+		return c.setType(&ast.CompositeLit{}, ty)
 	}
+	id := c.newIdent("nil", ty)
+	c.p.Uses[id] = nilObj
+	return id
+}
+
+func (c *funcContext) newConst(t types.Type, value exact.Value) ast.Expr {
+	id := &ast.Ident{}
+	c.p.Types[id] = types.TypeAndValue{Type: t, Value: value}
+	return id
 }
 
 func (c *funcContext) newVariable(name string) string {

--- a/compiler/utils.go
+++ b/compiler/utils.go
@@ -188,7 +188,7 @@ func (c *funcContext) zeroValue(ty types.Type) string {
 		return fmt.Sprintf("%s.zero()", c.typeName(ty))
 	case *types.Signature:
 		return "$throwNilPointerError"
-	case *types.Slice:
+	case *types.Slice, *types.Pointer, *types.Chan:
 		return fmt.Sprintf("%s.nil", c.typeName(ty))
 	case *types.Struct:
 		return fmt.Sprintf("new %s.ptr()", c.typeName(ty))
@@ -196,8 +196,9 @@ func (c *funcContext) zeroValue(ty types.Type) string {
 		return "false"
 	case *types.Interface:
 		return "$ifaceNil"
+	default:
+		panic(fmt.Sprintf("unexpected type: %T", t))
 	}
-	return fmt.Sprintf("%s.nil", c.typeName(ty))
 }
 
 func (c *funcContext) newVariable(name string) string {

--- a/tests/run.go
+++ b/tests/run.go
@@ -44,15 +44,6 @@ import (
 // TODO: Reduce these to zero or as close as possible.
 //
 var knownFails = map[string]failReason{
-	// GopherJS compiler panics.
-	"fixedbugs/issue9691.go": {category: compilerPanic, desc: `
-panic: Unhandled lhs type: *types.Map
-
-goroutine 1 [running]:
-github.com/gopherjs/gopherjs/compiler.(*funcContext).translateAssign(0xc820160210, 0xe00488, 0xc820105050, 0x632508, 0x2, 0xcb46b0, 0x908660, 0x0, 0x0, 0x0)
-	github.com/gopherjs/gopherjs/compiler/statements.go:785 +0x1931
-`},
-
 	"fixedbugs/bug243.go":    {desc: "TypeError: undefined is not a function"},
 	"fixedbugs/issue4495.go": {desc: "TypeError: Cannot read property 'm' of undefined"},
 

